pkg/discovery/dns/provider.go

```diff
@@ -0,0 +1,65 @@
+package dns
+
+import (
+	"context"
+	"strings"
+	"sync"
+)
+
+// Provider is a store for DNS resolved addresses. It provides a way to resolve addresses and obtain them.
+type Provider struct {
```

Comment:
This is again pretty much a cache. The only difference is that it applies some logic before storing addreses (in this case DNS lookup). I guess we can rename it to Cache, or is Provider more descriptive?

Reply:
For sure not cache, I would actually name it addressDiscover and move it to discovery package maybe? Because dns resolution is just some small part of this. The effect that query or rule or any user care is: 
Given:
- files for sd
- static entries
Return me addresses of store APIs.

Dns functionality is just additional here. That's my first impression (:

Reply:
Hmm, I think the DNS functionality is one of the only things this component is doing. It doesn't care where you get the addresses (fileSD or static flags), it only resolves a list of addresses using our custom prefix and stores the latest results (in a cache-like manner).